### ISSUE TITLE
Add info about binding when deploying applications

### ISSUE THREAD

**Issue by matthewhelmke** [juju/docs#1175]
I think this belongs in charms-deploying.md. Need to study it more closely.
## Binding Service Endpoints to Spaces

Binding means the "bound" endpoints will have IP addresses from
subnets that are part of the space the endpoint is bound to.

Use the optional '--bind' argument when deploying a service to
specify to which space individual charm endpoints should be bound. The
syntax for the '--bind' argument is a whitespace-separated list of
endpoint and space names, separated by "=".

When '--bind' is not specified, all endpoints will use the same
address, which is the host machine's preferred private address, as
returned by "unit-get private-address". This is backwards-compatible behaviour.

Additionally, a service-default space can be specified by omitting
the "<endpoint>=" prefix before the space name. This space will
be used for binding all endpoints that are not explicitly specified.

Examples:

```
juju deploy mysql --bind "db=database server=internal"
```

Bind "db" endpoint to an address part of the "database" space (i.e. the
address is coming from one of the "database" space's subnets in MAAS).

```
juju deploy wordpress --bind internal-apps
```

Bind _all_ endpoints of wordpress to the "internal-apps" space.

```
juju deploy haproxy --bind "url=public internal"
```

Bind "url" to "public", and all other endpoints to "internal".
## Binding Endpoints of Services Within a Bundle to Spaces

Bindings can be specified for services within a bundle the same way as
when deploying individual charms. The bundle YAML can include a section
called "bindings", defining the map of endpoint names to space names.

Example bundle.yaml excerpt:
    ...
    mysql:
        charm: cs:trusty/mysql-53
        num_units: 1
        constraints: mem=4G
        bindings:
            server: database
            cluster: internal
    ...

Deploying a bundle including a section like in the example above, is
equivalent to running:

```
juju deploy mysql --bind "server=database cluster=internal"
```

There is currently no way to declare a service-default space for all
endpoints in a bundle's bindings section. A workaround is to list all
endpoints explicitly.

**Comment by matthewhelmke** [comment]
Related.

## New Hook Command: network-get

When deploying a service with endpoint bindings specified, charm authors
can use the new "network-get" hook command to determine which address to
advertise for a given endpoint. This approach will eventually replace
"unit-get private-address" as well as various other ways to get the
address to use for a given unit.

There is currently a mandatory '--primary-address' argument to 'network-
get', which guarantees a single IP address to be returned.

Example (within a charm hook): 

    relation-ids cluster
    url:2

    network-get -r url:2 --primary-address
    10.20.30.23

(assuming the service was deployed with e.g. --bind url=internal, and
(10.20.30.0/24 is one of the subnets in that "internal" space).

**Comment by dimitern** [comment]
Please note the initial '-r <id>' syntax as described is no longer
supported. Now the first positional argument needs to be a binding name
(matching either one of those listed in extra-bindings or an existing
relation name). E.g. network-get admin-api --primary-address

On 06/10/2016 05:52 PM, Matthew Helmke wrote:
> Related.
> 
> 
>     New Hook Command: network-get
> 
> When deploying a service with endpoint bindings specified, charm authors
> can use the new "network-get" hook command to determine which address to
> advertise for a given endpoint. This approach will eventually replace
> "unit-get private-address" as well as various other ways to get the
> address to use for a given unit.
> 
> There is currently a mandatory '--primary-address' argument to 'network-
> get', which guarantees a single IP address to be returned.
> 
> Example (within a charm hook):
> 
> |relation-ids cluster url:2 network-get -r url:2 --primary-address
> 10.20.30.23 |
> 
> (assuming the service was deployed with e.g. --bind url=internal, and
> (10.20.30.0/24 is one of the subnets in that "internal" space).
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/juju/docs/issues/1175#issuecomment-225204121>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AAl10qMntNDRn59yhx2LadzLYkU_fOS1ks5qKXotgaJpZM4IzBrN>.
> 


-- 
Dimiter Naydenov <dimiter.naydenov@canonical.com>
Juju Core Sapphire team <http://juju.ubuntu.com>

**Comment by matthewhelmke** [comment]
Closed in #1341
